resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -215,13 +302,86 @@ void onResponse(final DnsQuestion question, AddressedEnvelope<DnsResponse, InetS
 
             // Retry with the next server if the server did not tell us that the domain does not exist.
             if (code != DnsResponseCode.NXDOMAIN) {
-                query(nameServerAddrs.next(), question, promise);
+                query(nameServerAddrStream, question, promise);
             }
         } finally {
             ReferenceCountUtil.safeRelease(envelope);
         }
     }
 
+    /**
+     * Handle redirects if needed and returns {@code true} if there was a redirect handled. If {@code true} is returned
+     * this method took ownership of the {@code promise} otherwise {@code false} is returned.
+     */
+    private boolean handleRedirect(
+            DnsQuestion question, AddressedEnvelope<DnsResponse, InetSocketAddress> envelope, Promise<T> promise) {
+        final DnsResponse res = envelope.content();
+
+        // Check if we have answers, if not this may be an non authority NS and so redirects must be handled.
+        if (res.count(DnsSection.ANSWER) == 0) {
+            AuthoritativeNameServerList serverNames = extractAuthoritativeNameServers(question.name(), res);
+
+            if (serverNames != null) {
+                List<InetSocketAddress> nameServers = new ArrayList<InetSocketAddress>(serverNames.size());
+                int additionalCount = res.count(DnsSection.ADDITIONAL);
+
+                for (int i = 0; i < additionalCount; i++) {
+                    final DnsRecord r = res.recordAt(DnsSection.ADDITIONAL, i);
+
+                    if ((r.type() == DnsRecordType.A && !parent.supportsARecords()) ||
+                            r.type() == DnsRecordType.AAAA && !parent.supportsAAAARecords()) {
+                        continue;
+                    }
+
+                    final String recordName = r.name();
+                    AuthoritativeNameServer authoritativeNameServer =
+                            serverNames.remove(recordName);
+
+                    if (authoritativeNameServer == null) {
+                        // Not a server we are interested in.
+                        continue;
+                    }
+
+                    InetAddress resolved = parseAddress(r, recordName);
+                    if (resolved == null) {
+                        // Could not parse it, move to the next.
+                        continue;
+                    }
+
+                    nameServers.add(new InetSocketAddress(resolved, parent.dnsRedirectPort(resolved)));
+                    addNameServerToCache(authoritativeNameServer, resolved, r.timeToLive());
+                }
+
+                if (nameServers.isEmpty()) {
+                    promise.tryFailure(
+                            new UnknownHostException("Unable to find correct name server for " + hostname));
+                } else {
+                    // Just use the same oder as we received the name servers for lookup.
```

Comment:
@Scottmitch yep... but maybe you are right and we should better just shuffle it everytime. Let me do this.